apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php

```diff
@@ -299,4 +303,55 @@ public function allow_searching_for_templates() {
 		// Setting this to `public` will allow it to be found in the search endpoint.
 		$post_type->public = true;
 	}
+
+	public function remove_template_components( $data, $postarr ) {
+		if ( 'page' !== $postarr['post_type'] ) {
+			return $data;
+		}
+
+		$template_id = get_post_meta( $postarr['ID'], '_wp_template_id', true );
+
+		if ( isset( $template_id ) && ! empty( $template_id ) ) {
+			$post_content = wp_unslash( $data['post_content'] );
+
+			preg_match_all(
+				'/<!-- wp:a8c\/template-wrapper[\s\S]+?wp:a8c\/template-wrapper -->/',
+				$post_content,
+				$template_parts
+			);
+
+			$before_content = str_replace(
+				array(
+					'<!-- wp:a8c/template-wrapper {"position":"before","align":"full"} -->',
+					'<!-- /wp:a8c/template-wrapper -->',
+				),
+				'',
+				$template_parts[0][0]
+			);
+			$after_content = str_replace(
+				array(
+					'<!-- wp:a8c/template-wrapper {"position":"after","align":"full"} -->',
+					'<!-- /wp:a8c/template-wrapper -->',
+				),
+				'',
+				$template_parts[0][1]
+			);
+
+			$updated_template_content = $before_content . "\n\n<!-- wp:a8c/post-content /-->\n\n" . $after_content;
+
+			wp_update_post( array(
+				'ID'           => $template_id,
+				'post_content' => $updated_template_content,
+			) );
+		}
+
+		$stripped_post_content = preg_replace(
+			'/<!-- wp:a8c\/template-wrapper[\s\S]+?wp:a8c\/template-wrapper -->/',
+			'',
+			$data['post_content']
+		);
```

Comment:
Posting here for your own reference, this may have bugs in it - ideally we wouldn't use `parse_blocks()` directly but given a few things noted in Slack it's a reasonable tradeoff here: namely we're trying to extract the blocks from the template but preserve the other blocks.

```php
// this is the easy part - this is what we write
function extract_template_contents( $post_content ) {
	$original_blocks = parse_blocks( $post_content );
	$blocks = [];

	foreach( $original_blocks as $block ) {
		if ( 'a8c/template-wrapper' !== $block['blockName'] ) {
			$blocks[] = serialize_block( $block );
		}

		// manually extract the inner contents of the template wrapper
		// if we have no plain HTML in there this becomes much easier
		//
		// foreach( $block['innerBlocks'] as $block ) {
		//   $blocks[] = serialize_block( $block );
		// }
		//
		$innerBlockIndex = 0;
		foreach ( $block['innerContent'] as $chunk ) {
			$blocks[] = null === $chunk
				? serialize_block( $block['innerBlocks'][$innerBlockIndex++] )
				: $chunk;
		}
	}

	return $blocks;
}

// this is the hard part - this should already be in there - sorry
function serialize_block( $block ) {
	$output = empty( $block['attrs'] )
		? "<!-- wp:{$block['blockName']} -->"
		: "<!-- wp:{$block['blockName']} " . wp_json_encode( $block['attrs'] ) . " -->";

	$innerBlockIndex = 0;
	foreach ( $block['innerContent'] as $chunk ) {
		$output .= null === $chunk
			? serialize_block( $block[$innerBlockIndex++] )
			: $chunk;
	}

	$output .= "<!-- /wp:{$block['blockName']} -->";

	return $output;
}
```

Reply:
there's a tiny little bug in there on JSON attribute serialization but I didn't want to include it in the example code because it would distract from the idea, plus the `serialize_block()` function should be provided by core - I'll get a patch together to add it

Reply:
https://core.trac.wordpress.org/ticket/47375